rusty-machine/src/linalg/matrix/impl_ops.rs

```diff
@@ -0,0 +1,648 @@
+///
+/// This will reuse allocated memory from `m`.
+impl<'a, T: Copy + $trt<T, Output=T>> $trt<Matrix<T>> for &'a Matrix<T> {
+    type Output = Matrix<T>;
+
+    fn $op(self, mut m: Matrix<T>) -> Matrix<T> {
+    	assert!(self.cols == m.cols, "Column dimensions do not agree.");
+        assert!(self.rows == m.rows, "Row dimensions do not agree.");
+
+        utils::in_place_vec_bin_op(&mut m.data, &self.data, |x,&y| {*x = (y).$op(*x)});
+        m
+    }
+}
+
+/// Performs elementwise
+#[doc=$doc]
+/// between two matrices.
+///
+/// This will reuse allocated memory from `self`.
+impl<'a, T: Copy + $trt<T, Output=T>> $trt<&'a Matrix<T>> for Matrix<T> {
+    type Output = Matrix<T>;
+
+    fn $op(mut self, m: &Matrix<T>) -> Matrix<T> {
+    	assert!(self.cols == m.cols, "Column dimensions do not agree.");
+        assert!(self.rows == m.rows, "Row dimensions do not agree.");
+
+        utils::in_place_vec_bin_op(&mut self.data, &m.data, |x,&y| {*x = (*x).$op(y)});
+        self
+    }
+}
+
+/// Performs elementwise
+#[doc=$doc]
+/// between two matrices.
+impl<'a, 'b, T: Copy + $trt<T, Output=T>> $trt<&'b Matrix<T>> for &'a Matrix<T> {
+    type Output = Matrix<T>;
+
+    fn $op(self, m: &Matrix<T>) -> Matrix<T> {
+        assert!(self.cols == m.cols, "Column dimensions do not agree.");
+        assert!(self.rows == m.rows, "Row dimensions do not agree.");
+
+        let new_data = utils::vec_bin_op(&self.data, &m.data, |x, y| { x.$op(y) });
+
+        Matrix {
+        	rows: self.rows,
+        	cols: self.cols,
+        	data: new_data,
+        }
+    }
+}
+    );
+);
+
+impl_bin_op_mat!(Add, add, "addition");
+impl_bin_op_mat!(Sub, sub, "subtraction");
+
+macro_rules! impl_neg_slice (
+    ($slice:ident) => (
+
+/// Gets negative of matrix slice.
+impl<T: Neg<Output = T> + Copy> Neg for $slice<T> {
+    type Output = Matrix<T>;
+
+    fn neg(self) -> Matrix<T> {
+        - &self
+    }
+}
+
+/// Gets negative of matrix slice.
+impl<'a, T: Neg<Output = T> + Copy> Neg for &'a $slice<T> {
+    type Output = Matrix<T>;
+
+    fn neg(self) -> Matrix<T> {
+        let new_data = self.iter().map(|v| -*v).collect();
+
+        Matrix {
+            cols: self.cols,
+            rows: self.rows,
+            data: new_data,
+        }
+    }
+}
+
+    );
+);
+
+impl_neg_slice!(MatrixSlice);
+impl_neg_slice!(MatrixSliceMut);
+
+/// Gets negative of matrix.
+impl<T: Neg<Output = T> + Copy> Neg for Matrix<T> {
+    type Output = Matrix<T>;
+
+    fn neg(mut self) -> Matrix<T> {
+        for val in self.data.iter_mut() {
+            *val = -*val;
+        }
+
+        self
+    }
+}
+
+/// Gets negative of matrix.
+impl<'a, T: Neg<Output = T> + Copy> Neg for &'a Matrix<T> {
+    type Output = Matrix<T>;
+
+    fn neg(self) -> Matrix<T> {
+        let new_data = self.data.iter().map(|v| -*v).collect();
+
+        Matrix {
+            cols: self.cols,
+            rows: self.rows,
+            data: new_data,
+        }
+    }
+}
```

Comment:
You can probably configure your error to automatically put a newline at end of files so that diff tools won't complain like this.